pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,835 @@
+	gardenletKubeconfigSecretName          = "gardenlet-kubeconfig"
+)
+
+func (a *actuator) getGardenletChartValues(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) (map[string]interface{}, error) {
+	var err error
+
+	// Determine deployment values
+	var deploymentValues map[string]interface{}
+	if deploymentValues, err = utils.ToValues(managedSeed.Spec.Gardenlet.Deployment); err != nil {
+		return nil, err
+	}
+	if managedSeed.Spec.Gardenlet.MergeParentConfig {
+		parentDeployment, err := getParentDeployment(a.imageVector, shoot)
+		if err != nil {
+			return nil, err
+		}
+		parentDeploymentValues, err := utils.ToValues(parentDeployment)
+		if err != nil {
+			return nil, err
+		}
+		deploymentValues = utils.MergeMaps(parentDeploymentValues, deploymentValues)
+	}
+
+	// Determine config values
+	var configValues map[string]interface{}
+	if configValues, err = utils.ToValues(managedSeed.Spec.Gardenlet.Config); err != nil {
+		return nil, err
+	}
+	if managedSeed.Spec.Gardenlet.MergeParentConfig {
+		parentConfig, err := getParentConfig(a.config)
+		if err != nil {
+			return nil, err
+		}
+		parentConfigValues, err := utils.ToValues(parentConfig)
+		if err != nil {
+			return nil, err
+		}
+		configValues = utils.MergeMaps(parentConfigValues, configValues)
+	}
+
+	// Marshal config values back to an object
+	var configObj *configv1alpha1.GardenletConfiguration
+	if err := utils.FromValues(configValues, &configObj); err != nil {
+		return nil, err
+	}
+
+	// If a bootstrap mechanism is specified, compute bootstrap kubeconfig and set gardenClientConnection values accordingly
+	// Otherwise, if kubeconfig path is specified in gardenClientConnection, read it and store its contents
+	var gccValues map[string]interface{}
+	if gccValues, err = utils.GetMapFromValues(configValues, "gardenClientConnection"); err != nil {
+		return nil, err
+	}
+	if managedSeed.Spec.Gardenlet.GardenConnectionBootstrap != nil {
+		// Compute bootstrap kubeconfig
+		address, _ := gccValues["gardenClusterAddress"].(*string)
+		caCert, _ := gccValues["gardenClusterCACert"].([]byte)
+		bootstrapKubeconfig, err := a.getBootstrapKubeconfig(ctx, managedSeed.Name, address, caCert, *managedSeed.Spec.Gardenlet.GardenConnectionBootstrap)
+		if err != nil {
+			return nil, err
+		}
+
+		// Set bootstrapKubeconfig and kubeconfigSecret and unset kubeconfig in gardenClientConnection values
+		gccValues = utils.InitValues(gccValues)
+		if bootstrapKubeconfig != "" {
+			gccValues["bootstrapKubeconfig"] = map[string]interface{}{
+				"name":       gardenletKubeconfigBootstrapSecretName,
+				"namespace":  v1beta1constants.GardenNamespace,
+				"kubeconfig": bootstrapKubeconfig,
+			}
+		}
+		gccValues["kubeconfigSecret"] = map[string]interface{}{
+			"name":      gardenletKubeconfigSecretName,
+			"namespace": v1beta1constants.GardenNamespace,
+		}
+		delete(gccValues, "kubeconfig")
```

Comment:
I would still keep it as it is to make it explicit and protect this code against potential changes in the Gardenlet logic you pointed to, which is in a completely unrelated part of the code. Regarding being used as a fallback - why would we even want this? I thought we wanted to make sure we go through the bootstrap process if `gardenConnectionBootstrap` is set, for security reasons?

Reply:
> I thought we wanted to make sure we go through the bootstrap process if gardenConnectionBootstrap is set, for security reasons

Yep, I think so. Being able to add a fallback kubeconfig is purely optional - if not set, the Gardenlet will fail. Sure we can discuss if that makes sense, just wanted to point out how it is currently designed.

> I would still keep it as it is to make it explicit and protect this code against potential changes in the Gardenlet logic you pointed 

There are configurations that a user has specified in the managed seed that are being deleted and not used for the Gardenlet helm chart. 
Instead of deleting it from the helm chart values, as a user, I would expect a validation error upon applying the ManagedSeed resource. And if not, I expect that all my configuration is taken into consideration.

**For validation (if we really want to be more restrictive for ManagedSeeds than is allowed in the GardenletConfiguration)**
`GardenConnectionBootstrap` is set 
  -  do not allow setting `managedSeed.Spec.Gardenlet.Config.seedClientConnection.kubeconfig` 
 
`managedSeed.Spec.Gardenlet.Config.seedClientConnection.kubeconfig` is set 
  - do not allow setting `managedSeed.Spec.Gardenlet.Config.seedClientConnection.bootstrapKubeconfig` and `kubeconfigSecret`

In addition, I think we need to make sure that we for one do not inherit invalid configuration via `MergeParentConfig` (hence you delete the kubeconfigSecret & bootstrapKubeconfig?) but at the same time do not override explicit configuration from the user. For instance, it is not possible atm to specify your own bootstrap kubeconfig in a secret in the Seed because it will always be deleted from the Gardenlets configuration.

Reply:
> Instead of deleting it from the helm chart values, as a user, I would expect a validation error upon applying the ManagedSeed resource. And if not, I expect that all my configuration is taken into consideration.

Why "instead of"? I think we need both. The validation would tell the user he configured something wrong. Explicitly deleting settings that are incompatible with other valid settings that the user did specify brings additional safety and protects from wrong settings in the parent.

> ... make sure that we for one do not inherit invalid configuration via MergeParentConfig (hence you delete the kubeconfigSecret & bootstrapKubeconfig?)

Exactly.

Regarding validation, I think the logic should be (I believe you mixed "garden" and "seed" above):

* If `GardenConnectionBootstrap` is set, `Config.gardenClientConnection.kubeconfig` must not be set (possibly also `Config.gardenClientConnection.bootstrapKubeconfig` and `Config.gardenClientConnection.kubeconfigSecret` if these remain hardcoded, see the other discussion)
* Conversely, if `GardenConnectionBootstrap` is not set, `Config.gardenClientConnection.kubeconfig` must be set if merging with parent is disabled.
* If `SeedConnection` is set, `Config.seedClientConnection.kubeconfig` must not be set.
* If `SeedConnection` is not set, `Config.seedClientConnection.kubeconfig` must be set even if merging with parent is enabled, since it's virtually guaranteed that the parent configuration would be wrong.

Reply:
We decided to cleanup `gardenConnectionClient.kubeconfig` explicitly if `GardenConnectionBootstrap`, as is already the case. The validation logic described in my last comment should also be implemented (except for `SeedConnection` that we decided to remove.